pkg/querier/vcs/client/github.go

```diff
@@ -0,0 +1,157 @@
+		ClientSecret: githubAppClientSecret,
+		Endpoint:     o2endpoints.GitHub,
+	}, nil
+}
+
+// GithubClient returns a github client for the given request headers.
+func GithubClient(ctx context.Context, requestHeaders http.Header) (*githubClient, error) {
+	auth, err := githubOAuth()
+	if err != nil {
+		return nil, err
+	}
+	cookie, err := (&http.Request{Header: requestHeaders}).Cookie(gitHubCookieName)
+	if err != nil {
+		return nil, err
+	}
+	token, err := decryptToken(cookie.Value, githubSessionSecret)
+	if err != nil {
+		return nil, unAuthorizeError(err, cookie)
+	}
+	return &githubClient{
+		client: github.NewClient(auth.Client(ctx, token)),
+	}, nil
+}
+
+func unAuthorizeError(err error, cookie *http.Cookie) error {
+	connectErr := connect.NewError(
+		connect.CodeUnauthenticated,
+		err,
+	)
+	cookie.Value = ""
+	cookie.MaxAge = -1
+	connectErr.Meta().Set("Set-Cookie", cookie.String())
+	return connectErr
+}
+
+func AuthorizeGithub(ctx context.Context, authorizationCode string) (string, error) {
+	auth, err := githubOAuth()
+	if err != nil {
+		return "", err
+	}
+	token, err := auth.Exchange(ctx, authorizationCode)
+	if err != nil {
+		return "", err
+	}
+	cookieValue, err := encryptToken(token, githubSessionSecret)
+	if err != nil {
+		return "", err
+	}
+	// Sets a cookie with the encrypted token.
+	// Only the server can decrypt the cookie.
+	cookie := http.Cookie{
+		Name:  gitHubCookieName,
+		Value: cookieValue,
+		// Refresh expiry is 6 months based on github docs
+		Expires:  time.Now().Add(15811200 * time.Second),
+		HttpOnly: false,
+		Secure:   true,
+		SameSite: http.SameSiteLaxMode,
+	}
+	// responseHeaders.Add("Set-Cookie", cookie.String())
+	return cookie.String(), nil
+}
+
+type githubClient struct {
+	client *github.Client
+}
+
+func (gh *githubClient) GetCommit(ctx context.Context, owner, repo, ref string) (*vcsv1.GetCommitResponse, error) {
+	commit, _, err := gh.client.Repositories.GetCommit(ctx, owner, repo, ref, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &vcsv1.GetCommitResponse{
+		Sha:     toString(commit.SHA),
+		Message: toString(commit.Commit.Message),
+		Author: &vcsv1.CommitAuthor{
+			Login:     toString(commit.Author.Login),
+			AvatarURL: toString(commit.Author.AvatarURL),
+		},
+		Date: commit.Commit.Author.Date.Format(time.RFC3339),
+	}, nil
+}
+
+func (gh *githubClient) GetFile(ctx context.Context, req FileRequest) (File, error) {
+	// We could abstract away git provider using git protocol
+	// git clone https://x-access-token:<token>@github.com/owner/repo.git
+	// For now we use the github client.
+	file, _, _, err := gh.client.Repositories.GetContents(ctx, req.Owner, req.Repo, req.Path, &github.RepositoryContentGetOptions{Ref: req.Ref})
+	if err != nil {
+		var githubErr *github.ErrorResponse
+		if ok := errors.As(err, &githubErr); ok && githubErr.Response.StatusCode == http.StatusNotFound {
+			return File{}, fmt.Errorf("%w: %s", ErrNotFound, err)
+		}
+		return File{}, err
```

Comment:
Will this propagate errors correctly? I.e. if github sends back a 422 do we bubble that back to the frontend?

Reply:
No for other errors and I'm not sure we should. TBD

Reply:
The content will though contains more info.